core/src/main/java/org/mule/TransformationService.java

```diff
@@ -429,19 +427,29 @@ private void transformMessage(MuleMessage message, MuleEvent event, Transformer
                 // instance
                 synchronized (message)
                 {
-                    MuleMessage resultMessage = (MuleMessage) result;
-                    message.setPayload(resultMessage.getPayload(), resultMessage.getDataType());
-                    message.setOriginalPayload(resultMessage.getOriginalPayload());
-                    message.copyMessageProperties(resultMessage);
-                    message.copyAttachments(resultMessage);
+                    MuleMessage transformResult = (MuleMessage) result;
+                    resultMessage = new DefaultMuleMessage(transformResult.getPayload(), transformResult, muleContext, transformResult.getDataType());
+                    resultMessage.setOriginalPayload(transformResult.getOriginalPayload());
+                    resultMessage.copyMessageProperties(transformResult);
+                    resultMessage.copyAttachments(transformResult);
                 }
             }
         }
         else
         {
             final DataType<?> mergedDataType = mergeDataType(message, message.getDataType(), transformer.getReturnDataType());
-            message.setPayload(result, mergedDataType);
+            resultMessage = new DefaultMuleMessage(result, message, muleContext, mergedDataType);
         }
+
+        if (message.getOriginalPayload() == null && muleContext.getConfiguration().isCacheMessageOriginalPayload())
+        {
+            resultMessage.setOriginalPayload(message.getPayload());
+        }
+
+        // Update the RequestContext with the result of the transformation.
+        RequestContext.internalRewriteEvent(resultMessage, false);
```

Comment:
This was always here, I just moved it down to after construction of new message instance.